### ISSUE TITLE
Improve spawn module error messages

### ISSUE THREAD

**Issue by techee** [geany/geany#541]
When playing with Geany under Windows I noticed that the message saying grep couldn't be executed is in 1.25:

search_find_in_files: spawn_with_callbacks() failed: CreateProcess() failed: The system cannot find the file specified.

First it doesn't tell the name of the executable which failed to start, second it IMO contains too many internal details which may be confusing for users. The previous message was

Cannot execute grep tool 'grep'; check the path setting in Preferences.

which is much more user-friendly. This one is clearly tuned for this particular case but a generic message could look like

Failed to execute '%s', executable_name

and it would be much clearer for users what's going on.

**Comment by zhekov** [comment]
The last part, "CreateProcess() failed: The system cannot find the file specified.", is generated by spawn_*(). Normally it would be the error text only, but spawn_*() is a complex call, with several other (though unlikely) points of failure, thus the function name.

As of the executable name, only the caller really knows whether it's locale or UTF-8 (having a locale names does not guarantee that all names are locale, that's a basic error in Geany). So spawn_*() is encoding-agnostic, but a GError message must be in UTF-8, and thus the executable is not included.

That is not to say that I'm against more sensible messages - we even discussed it while polishing spawn, but could not reach any sensible conclusion. Perhaps it would be best to display user-friendly messages based on the most likely error, and put the full details in the Status tab, or Debug -> Messages.

**Comment by elextr** [comment]
Really libraries should not display errors, they should return error messages to the caller so the caller can decorate it with whatever more information they have.

So spawn_* would return the generic "Failed to find the executable" and the grep caller would combine to make a message like "Failed to execute 'grep': Failed to find executable".

**Comment by zhekov** [comment]
The spawn calls display nothing, of course, they only return a GError (and even know nothing about gtk+, only glib is used).

Since I haven't seen any internally used win~1 functions except CreateProcess (== exec) to produce any errors, removing the function name, except when compiling the test program, seems acceptable to me.

One unpleasant problem is that, if the command line can not be parsed properly, glib often includes it's text in the GError message, along with the translated error description. If the command line is in locale, it won't be displayed properly (Help -> Debug Messages will even terminate Geany with an assertion, as in PR #658). But the (entire) message can not be converted to utf-8 by the caller, because the translated text will break.

Maybe I should refrain from mimicking this glib behavior under win~1, and even replace the offending glib messages under *nix? It's possible, based on their GError codes.

**Comment by techee** [comment]
@zhekov In my opinion, the whole prefix of the message that should be removed, i.e.:

search_find_in_files: spawn_with_callbacks() failed: CreateProcess() failed: 

This is something the users shouldn't see when GError message is displayed - it's just something that might be useful for debugging (might be fine to put it to the debugging messages).

Regarding the grep message, I think it wasn't a good idea to replace the original error message:

		ui_set_statusbar(TRUE, _("Cannot execute grep tool '%s';"
			" check the path setting in Preferences."), tool_prefs.grep_cmd);

with

	if (command_grep == NULL)
		command_line = g_strdup_printf("%s %s --", tool_prefs.grep_cmd, opts);
	else
	{
		command_line = g_strdup_printf("\"%s\" %s --", command_grep, opts);

The original message didn't use the GError at all and I think in this case it's the right thing to do - we know that the grep command failed to execute so we tell the user plus the info what needs to be changed.

I didn't go through all the uses of the new spawn module but I can imagine such a thing happens on other places too - from the special case where the spawn() is called we can probably create a more useful error message than what the spawn() itself can generate.

**Comment by techee** [comment]
Just went through the code quickly and one more useful message that got removed in

690cb922be902f023881d455ae0c0a87d1c62170

is

			ui_set_statusbar(TRUE,
				_("Could not parse terminal command \"%s\" "
					"(check Terminal tool setting in Preferences)"), tool_prefs.term_cmd);

(The other message I was referring to previously was removed in 74171cca5026e915e68b55f420b986b266ddcf7f)

Apart from these messages it should be OK (together with removal of the debugging messages from GError).

**Comment by zhekov** [comment]
@techee 

> In my opinion, the whole prefix of the message that should be removed [...]

I can remove the "FunctionName() failed:" from the spawn module. The other prefixes, if any, are added by the callers.

> ui_set_statusbar(TRUE, _("Cannot execute grep tool '%s';"
>        " check the path setting in Preferences.")

This is indeed user friendly, but tells us absolutely nothing about why grep failed. My suggestion above was to maybe display a user friendly message, but also put the full text in Status or Debug -> Messages.

If we are going to display something like this, with no information from the actual GError message, there would be no reason to touch the spawn module...

(BTW, this particular message was not correct, unless you read "cannot execute" very widely: it was produced if g_find_program_in_path() failed, and the execution itself was about 60 lines below, with a completely different message.)

> _("Could not parse terminal command \"%s\" "
>                "(check Terminal tool setting in Preferences)")

spawn_async() parses the command line and reports any errors in more detail. It also rejects insecure program names under Windows, which is not really a parsing error.

TL;DR:

Now, if I also remove the parsed command text from the spawn GError, leaving only the error description text (my 2nd suggestion), these may be unified and become:

Cannot execute grep tool "%s": <error>. Check the path setting in Preferences.
Cannot execute terminal command "%s": <error>. Check the path setting in Preferences.

For example:

Cannot execute grep tool "c:\msys32\usr\bin\grep.exe": The system cannot find the file specified. Check the path setting in Preferences.

Cannot execute terminal command "": Text was empty (or contained only whitespace). Check the path setting in Preferences.

Seems pretty reasonable to me.

**Comment by techee** [comment]
@zhekov Sounds good to me the way you suggest.

**Comment by zhekov** [comment]
It's similar to what @elextr suggested too, with a slightly different prefix and an user-friendly suffix, so I'll implement it when I have some free time.

The build commands are different though, I'll leave them as is for now.

**Comment by zhekov** [comment]
Finished, please test. As noted above, the build command (start and kill) messages are different, so I'm leaving them as is, unless we have clarity on their texts.
